### PR TITLE
Liger iOS Xcode Project Generator

### PR DESCRIPTION
**Pull request by jaredbro**
Initial generator for a base Liger iOS Xcode project.  This does not yet include the CoCoPods integration.

Fixes issues #7 and #8

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 node_modules/
+.DS_Store
```
Comment (ligerjohn) [inline .gitignore]: I wonder if there's more things we want to add in here, since it looks like the generated file is a bit, weak.

[Examples of gitignores](https://github.com/github/gitignore)

File: generators/ios/index.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+module.exports = yeoman.generators.Base.extend({
+  initializing: function() {
+    this.argument('projectName', {
+      required: false,
+      type: String,
+      desc: projectNameDesc
+    });
+
+    this.argument('packageName', {
```
Comment (ligerjohn) [inline generators/ios/index.js]: * Could we do a simple check to make sure that there aren't any wonky characters or other "Bad" things in here?
* The default for this value in the UI is "the com.company.etc thing you have set as default in Xcode" + name of app. Any chance we could gleam that for a good "default" if it's set? I've got it set for the company I'm working for for instance.

File: generators/ios/index.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
+'use strict';
+var yeoman = require('yeoman-generator');
+var chalk = require('chalk');
+var yosay = require('yosay');
+var path = require('path');
+
+var projectNameDesc = 'The Application Name';
+var packageNameDesc = 'A Company identifier following reverse-domain style convention (e.g. com.example.YourCompanyName)';
+var basePath = 'ios';
```
Comment (ligerjohn) [inline generators/ios/index.js]: I've traditionally gone with iOS and Android, but that is in no way needed (or maybe even desired).

Just wanted to make a note and say that we should default this to 'ios' and 'android' from now on.

File: generators/ios/index.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  prompting: function() {
+    if (this.projectName !== undefined && this.packageName !== undefined) {
+      return;
+    };
+
+    var done = this.async();
+
+        // Have Yeoman greet the user.
+    this.log(yosay(
+      'Welcome to the ' + chalk.red('LigerMobile iOS') + ' app generator!'
```
Comment (ligerjohn) [inline generators/ios/index.js]: Can't remember what we did for the other ones, but probably should avoid "red" as it's kinda "STOP, WARNING, ERROR, ACHTUNG!" :) We should fix the other ones as well if there are reds in there.

File: generators/ios/index.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    ));
+
+    var prompts = [
+      {
+          name: 'projectName',
+          message: 'What is the name of your app? (Spaces aren\'t allowed)',
```
Comment (ligerjohn) [inline generators/ios/index.js]: Is there a way we can enforce this? Maybe the prompt has an option?

(The motivation is to avoid letting a user do something, just to turn around and say "you can't do that!". Better to not allow faulty input. Second best is to do turn around. Letting it generate anyways is probably a bad idea.)

File: generators/ios/index.js [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    // Setup the options object for handling templates
+    this.templateOptions = { projectName: this.projectName, packageName: this.packageName };
+
+    this._writingProject(this.projectName, this.packageName);
+    this._writingXcodeProject(this.projectName, this.packageName);
+    this._writingTestProject(this.projectName, this.pakcage_name);
+  },
+
+  _writingProject: function() {
+    this.log('Creating app project ' + this.projectName + '...');
```
Comment (ligerjohn) [inline generators/ios/index.js]: Use '…' instead of '...'

(option + ; creates the ellipsis character rather than three '.' after each other)

File: generators/ios/index.js [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+    this.fs.copyTpl(
+      this.templatePath('project/__PROJECT_NAME__/AppDelegate.*'),
+      this.destinationPath(path.join(basePath, this.projectName)),
+      this.templateOptions
+    );
+
+    this.fs.copyTpl(
+      this.templatePath('project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib'),
+      this.destinationPath(path.join(basePath, this.projectName, 'Base.lproj', 'LaunchScreen.xib')),
```
Comment (ligerjohn) [inline generators/ios/index.js]: We should create a nice default liger launch screen xib to include. That way we can have a nice pdf (automatically generates PNGs from it) of the Liger logo in the launch screen. Plus we should add in the name of the app (unless we already do).

This might work better as an added issue rather than a change for this PR. In that case just comment with the new issue number.

File: generators/ios/index.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    this.fs.copyTpl(
+      this.templatePath('project/__PROJECT_NAME__/main.m'),
+      this.destinationPath(path.join(basePath, this.projectName, 'main.m')),
+      this.templateOptions
+    );
+
+    this.fs.copyTpl(
+      this.templatePath('project/__PROJECT_NAME__/Info.plist'),
+      this.destinationPath(path.join(basePath, this.projectName, 'Info.plist')),
```
Comment (ligerjohn) [inline generators/ios/index.js]: How do we make sure this is created in the correct locale or are we doing "everyone uses English"?

TL;DR iOS/OSX apps have a default locale, which dictates what locale is used for translations and which locale gets automatically created. All apps have at least one. A number of files are included in this. Usually quite few (2 or so) in a new project.

File: generators/ios/templates/pods/Podfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+# Uncomment this line to define a global platform for your project
+# platform :ios, '6.0'
```
Comment (ligerjohn) [inline generators/ios/templates/pods/Podfile]: Let's run with '6.0' as the min version as ligermobile itself doesn't really support anything below that.

[Looked it up](https://github.com/CocoaPods/Specs/blob/master/Specs/LigerMobile/0.2.1/LigerMobile.podspec.json) and it seems like we bottom out at 6.1 at the moment. So let's go with 6.1.

File: generators/ios/templates/project/__PROJECT_NAME__/AppDelegate.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+//  AppDelegate.h
+//  <%= projectName %>
+//
+//  Created by ___FULLUSERNAME___ on ___DATE___.
+//  Copyright ___ORGANIZATIONNAME___ ___YEAR___. All rights reserved.
+//
+
+#import <UIKit/UIKit.h>
+
+@interface AppDelegate : UIResponder <UIApplicationDelegate>
```
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/AppDelegate.h]: If you set the default "preamble" (can't remember the word) in Xcode all classes created will have that set. So it would be LMJAppDelegate for me for instance (LMJ as in LigerMobileJohn). Would be nice if we could get that in as well. It's generally not advisable to write apps without them as you can get name-clashes.

File: generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        <capability name="Constraints with non-1.0 multipliers" minToolsVersion="5.1"/>
+    </dependencies>
+    <objects>
+        <placeholder placeholderIdentifier="IBFilesOwner" id="-1" userLabel="File's Owner"/>
+        <placeholder placeholderIdentifier="IBFirstResponder" id="-2" customClass="UIResponder"/>
+        <view contentMode="scaleToFill" id="iN0-l3-epB">
+            <rect key="frame" x="0.0" y="0.0" width="480" height="480"/>
+            <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
+            <subviews>
+                <label opaque="NO" clipsSubviews="YES" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="  Copyright (c) 2015 Liger. All rights reserved." textAlignment="center" lineBreakMode="tailTruncation" baselineAdjustment="alignBaselines" minimumFontSize="9" translatesAutoresizingMaskIntoConstraints="NO" id="8ie-xW-0ye">
```
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib]: We should replace the (C) here with something relevant. The app ain't ours :)

File: generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                    <rect key="frame" x="20" y="439" width="441" height="21"/>
+                    <fontDescription key="fontDescription" type="system" pointSize="17"/>
+                    <color key="textColor" cocoaTouchSystemColor="darkTextColor"/>
+                    <nil key="highlightedColor"/>
+                </label>
+                <label opaque="NO" clipsSubviews="YES" userInteractionEnabled="NO" contentMode="left" horizontalHuggingPriority="251" verticalHuggingPriority="251" text="<%= projectName %>" textAlignment="center" lineBreakMode="middleTruncation" baselineAdjustment="alignBaselines" minimumFontSize="18" translatesAutoresizingMaskIntoConstraints="NO" id="kId-c2-rCX">
```
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib]: Ignore the comment from ```index.js```, looking good.

File: generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json [hunk 13]
```diff
@@ -0,0 +1,68 @@
+{
```
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json]: We should include the LM app icons so those are the defaults for all newly generated apps.

File: generators/ios/templates/project/__PROJECT_NAME__/Info.plist [hunk 14]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Info.plist]: Need to modify my comment in ```index.js```, this file ins't actually localized, it's another similar file. Grab hold of me if you want to chat localized files.

File: generators/ios/index.js (hunk 2 above)
Comment (ligerjohn) [inline generators/ios/index.js]: Realized that it shouldn't be + name, that's taken care of by the Info.plist file.

When I think of it, shouldn't that file be named Info-NameOfApp.plist?

File: generators/ios/index.js (hunk 8 above)
Comment (ligerjohn) [inline generators/ios/index.js]: Wrong file, but the idea still stands. I just didn't see those files here.

File: generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json (hunk 13 above)
Comment (jaredbro) [inline generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json]: That would be awesome.  Can you send me an example of an app that includes those for LM?

File: generators/ios/index.js (hunk 7 above)
Comment (ligerjohn) [inline generators/ios/index.js]: We already added in the name:) We should fix the © notice though.

**Comment by ligerjohn** [comment]
Done reviewing.

File: .gitignore (hunk 1 above)
Comment (jaredbro) [inline .gitignore]: Left this alone since I am not sure if there is anything else we want to add for sure.  This is just the generators .gitignore, but so we will have js, objective-c, java, etc in here eventually.  Lots of different things and don't want to "guess" at this point.

File: generators/ios/index.js (hunk 3 above)
Comment (jaredbro) [inline generators/ios/index.js]: Made the decisions with @ligerjohn to keep it lowercase

File: generators/ios/index.js (hunk 4 above)
Comment (jaredbro) [inline generators/ios/index.js]: Add issue to track this https://github.com/reachlocal/generator-ligermobile/issues/13

File: generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib (hunk 11 above)
Comment (jaredbro) [inline generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib]: Ok, what should this one be?
Technically this file is ours, we generated it from our own template.  But, I'm happy to empty it out as blank for now.

Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Base.lproj/LaunchScreen.xib]: Copyright (C) <%= some_variable_that_holds_the_current_year %> Your Company. All rights reserved.

Unless we have a company name we can use already in there.

Is the file after generation still (C) whoever wrote it or does it imply transfer of (c)-ship? I totally have no idea about the legal aspects here:)

That string shows up on the launch screen when the app starts btw. We could leave it as is and when we design a proper "default liger app" xib we just don't include the (c). It's very optional after all, it's just the standard template Xcode creates. It's meant to be changed anyways.

File: .gitignore (hunk 1 above)
Comment (ligerjohn) [inline .gitignore]: I was thinking of a node based gitignore. Since we won't actually use other langs for anything but templates we don't have to worry about anything being generated by anything but node itself.

File: generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json (hunk 13 above)
Comment (ligerjohn) [inline generators/ios/templates/project/__PROJECT_NAME__/Images.xcassets/AppIcon.appiconset/Contents.json]: [I'd look here](https://github.com/reachlocal/liger-ios/tree/master/LigerResources)

We should consider removing them from liger-ios as they probably belong here in the generator anyways.

**Comment by ligerjohn** [comment]
Change CoCo Pods to CocoaPods and we have a :+1:
